src/System.IO.Compression/src/System/IO/Compression/DeflateZLib/DeflateStream.cs

```diff
@@ -988,10 +990,11 @@ public override async Task WriteAsync(byte[] buffer, int offset, int count, Canc
                     {
                         await _destination.WriteAsync(new ReadOnlyMemory<byte>(_arrayPoolBuffer, 0, bytesRead), cancellationToken).ConfigureAwait(false);
                     }
-                    else
+                    else if (_deflateStream._inflater.NeedsInput())
```

Comment:
Why is this an `else` but it's not in the similar earlier cases?

Reply:
I actually need to look more closely at the states here. I had previously left all of these out of else and it caused one unrelated unit test to break. I need to debug to understand what’s happening there.  It may make more sense to use logic similar to what’s in Read.

Reply:
It turns out we don't expose enough state from the ZLIB operation to make this check based solely on inflater state.  The inflater can be in the "NeedsInput" state even if there is still more data to pull out from the native side, and none of the properties exposed on the inflater/ZLibStream today represent this case.  Instead we should check, only when the inflater returns 0 data if it did so because it ran out of input, or if it merely reached the end of one GZip payload., and can resume inflating the next payload.

IOW: this needs to always be else.